components/automate-ui/src/app/modules/infra-proxy/treetable/component/treetable.component.ts

```diff
@@ -0,0 +1,102 @@
+import { ValidatorService } from '../services/validator/validator.service';
+import { ConverterService } from '../services/converter/converter.service';
+import { defaultOptions } from '../default.options';
+import { flatMap, defaults } from 'lodash';
+import { Subject } from 'rxjs';
+import * as O from 'fp-ts/lib/Option';
+import { pipe } from 'fp-ts/lib/pipeable';
+
+@Component({
+  selector: 'app-treetable',
+  templateUrl: './treetable.component.html',
+  styleUrls: ['./treetable.component.scss']
+})
+export class TreetableComponent<T> implements OnInit, OnChanges {
+  @Input() tree: Node<T> | Node<T>[];
+  @Input() options: Options<T> = {};
+  @Output() nodeClicked: Subject<TreeTableNode<T>> = new Subject();
+  private searchableTree: SearchableNode<T>[];
+  private treeTable: TreeTableNode<T>[];
+  displayedColumns: string[];
+  dataSource: MatTableDataSource<TreeTableNode<T>>;
+  constructor(
+    private treeService: TreeService,
+    private validatorService: ValidatorService,
+    private converterService: ConverterService
+  ) { }
+
+  ngOnInit() {
+    this.tree = Array.isArray(this.tree) ? this.tree : [this.tree];
+    this.options = this.parseOptions(defaultOptions);
+    const customOrderValidator =
+    this.validatorService.validateCustomOrder(this.tree[0], this.options.customColumnOrder);
+    if (this.options.customColumnOrder && !customOrderValidator.valid) {
+      throw new Error(`
+        Properties ${customOrderValidator.xor.map(x => `'${x}'`).join(', ')} incorrect or missing in customColumnOrder`
+      );
+    }
+    this.displayedColumns = this.options.customColumnOrder
+      ? this.options.customColumnOrder
+      : this.extractNodeProps(this.tree[0]);
+    this.searchableTree = this.tree.map(t => this.converterService.toSearchableTree(t));
+    const treeTableTree = this.searchableTree.map(st => this.converterService.toTreeTableTree(st));
+    this.treeTable = flatMap(treeTableTree, this.treeService.flatten);
+    this.dataSource = this.generateDataSource();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes.tree.isFirstChange()) {
+      return;
+    }
+    this.tree = Array.isArray(this.tree) ? this.tree : [this.tree];
+    this.searchableTree = this.tree.map(t => this.converterService.toSearchableTree(t));
+    const treeTableTree = this.searchableTree.map(st => this.converterService.toTreeTableTree(st));
+    this.treeTable = flatMap(treeTableTree, this.treeService.flatten);
+    this.dataSource = this.generateDataSource();
+  }
+
+  extractNodeProps(tree: Node<T> & { value: { [k: string]: any } }): string[] {
+    return Object.keys(tree.value).filter(x => typeof tree.value[x] !== 'object');
+  }
+
+  generateDataSource(): MatTableDataSource<TreeTableNode<T>> {
+    return new MatTableDataSource(this.treeTable.filter(x => x.isVisible));
+  }
+
+  formatIndentation(node: TreeTableNode<T>, step: number = 5): string {
+    return '&nbsp;'.repeat(node.depth * step);
```

Comment:
Having to fiddle with HTML spacing like this is concerning. Is there no way to do this with CSS classes? How many levels might there be?

Reply:
for now we are assuming 5.

Reply:
So might this be done with CSS classes instead?

Reply:
@msorens 
I have removed `&nbsp` and added dynamic CSS function for this.

Reply:
regarding `level` I mentioned 5, but 5 is just a number to assign the steps(to set the spaces). 
we have `n` number of level and level depends(is dynamic) on the depth of hash value.